man/unipagecount.1
.TH UNIPAGECOUNT 1 "2007 Dec 31"
.SH NAME
unipagecount \- Count the assigned code points in a GNU Unifont .hex file
.SH SYNOPSIS
\fBunipagecount \fP[\-P\fIplane\fP] [\-p\fIpagenum\fP] [\-h|\-l]
.SH DESCRIPTION
.B unipagecount
reads a GNU Unifont .hex file from STDIN and prints a 16 by 16 grid of the
number of defined code points in each 256 character block within a Unicode
plane to STDOUT.  Code points proceed from left to right, then top to bottom.
In all planes, code points U+*FFFE and U+*FFFF are not expected in the input
hex file; they are reserved and always counted as being present in a plane.
.SH OPTIONS
.TP 6
.BR \-P
Select a Unicode plane, from 0 through 16, inclusive.  If not specified,
.B unipagecount
defaults to Plane 0 (the Basic Multilingual Plane).
.TP
.BR \-p
Just print information on one 256 code point "page" rather than
the entire Basic Multilingual Plane.  This prints a 16 by 16 table with
an asterisk in every code point that has an assigned glyph.
.TP
.BR \-h
Print an HTML table with color-coded cell background colors
instead of a plain text table.
.TP
.BR \-l
[The letter "l"]: Print hyperlinks to font bitmaps in the HTML table.
To create the bitmaps themselves, use the
.B unihex2bmp
program.  The bitmaps are assumed to be in the directory "bmp/".
.SH FILES
*.hex GNU Unifont font files
.SH SEE ALSO
.BR bdfimplode (1),
.BR hex2bdf (1),
.BR hex2otf (1),
.BR hex2sfd (1),
.BR hexbraille (1),
.BR hexdraw (1),
.BR hexkinya (1),
.BR hexmerge (1),
.BR johab2syllables (1),
.BR johab2ucs2 (1),
.BR unibdf2hex (1),
.BR unibmp2hex (1),
.BR unibmpbump (1),
.BR unicoverage (1),
.BR unidup (1),
.BR unifont (5),
.BR unifont-johab631 (5),
.BR unifont-viewer (1),
.BR unifont1per (1),
.BR unifontchojung (1),
.BR unifontksx (1),
.BR unifontpic (1),
.BR unigen-hangul (1),
.BR unigencircles (1),
.BR unigenwidth (1),
.BR unihex2bmp (1),
.BR unihex2png (1),
.BR unihexfill (1),
.BR unihexgen (1),
.BR unihexpose (1),
.BR unihexrotate (1),
.BR unijohab2html (1),
.BR unipng2hex (1)
.SH AUTHOR
.B unipagecount
was written by Paul Hardy.
.SH LICENSE
.B unipagecount
is Copyright \(co 2007, 2014 Paul Hardy.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH BUGS
No known real bugs exist, except that this software does not perform
extensive error checking on its input files.  If they're not in the
format of the original GNU Unifont .hex file, all bets are off.


NEWS
2024-02-24: Release 15.1.05<F2>
  * Added almost 600 CJK ideographs, including:
    - The entire CJK Unified Ideographs Extension D range of 222 glyphs
    - 335 commonly used Cantonese ideographs in Plane 2 and Plane 3.
    - Several updates to CJK Unified Ideographs Extension I.
  * All Hangul glyphs now have a consistent look with the Hangul Syllables,
    with new glyyphs for:
    - U+1100..U+11FF Hangul Jamo
    - U+3130..U+318F Hangul Compatibility Jamo
    - U+A960..U+A97F Hangul Jamo Extended-A
    - U+D7B0..U+D7FF Hangul Jamo Extended-B.
  * Several other glyph modifications; see the ChangeLog file for details.

2023-10-29: Release 15.1.04
  * Added larger Plane 2 supersets to default and Japanese font files.

2023-10-21: Release 15.1.03
  * Boris Zhang and Yzy32767 contributed CJK Unified Ideographs
    Extension I glyphs (U+2EBF0..U+2EE5D).
  * 湖 远星 contributed 14 glyph updates to CJK Unified Ideographs
    Extension B and Extension C: U+2A6D7..U+2A6DF and
    U+2B735..U+2B739.

2023-09-21: Release 15.1.02
   * 湖 远星 adjusted 46 glyphs in the Wen Quan Yi range, U+2F00..U+9FFF, and
     contributed CJK Unified Ideographs Extension G glyphs U+30000..U+3017F.
   * Added new man pages to MAN1PAGES & MAN5PAGES definitions for installation.

2023-09-12: Release 15.1.01
   * Glyphs updated for Unicode 15.1.0 release.
   * OpenType fonts are now built by default, replacing TrueType fonts.
     TrueType fonts can be built manually by invoking "make truetype"
     in the font directory.
   * New Hangul Johab 6/3/1 encoding devised by Ho-Seok Ee for building
     the Hangul Syllables block (U+AC00..U+D7A3), with new supporting
     programs johab2syllables, unigen-hangul, and unijohab2html.  This
     encoding replaces Unifont's original Hanterm-derived encoding.
     New source files hangul.h and unihangul-support.c provide support
     for rendering syllable bitmaps with all Unicode Hangul letters,
     including all ancient Hangul, through the new unigen-hangul program.
     These functions also can support user-written Hangul syllable rendering
     programs, for example on embedded systems using graphics displays.
   * Added three hexadecimal digit representation CSUR/UCSUR scripts:
     - U+EBE0..U+EBEF: Boby Lapointe's "bibi-binary" notation
     - U+EBF0..U+EBFF: Bruce Alan Martin's bit location notation
     - U+ECF0..U+ECFF: Ronald O. Whitaker's triangular notation.
   * Modified hexdraw to support bitmap manipulation of up to 64 rows
     by 128 columns for 64-by-128 graphics display screen drawing support.
     Added the unihexpose program to transpose Unifont glyphs, for use
     with graphics display controller chips that encode each byte as
     8 rows by 1 column, left to right.
   * Other minor updates; see ChangeLog for details.

2023-06-04: Release 15.0.06
   * Further adjustments to Hangul glyphs in the Basic Multilingual Plane.

2023-06-03: Release 15.0.05
   * New Hangul Syllables block (U+AC00..U+D7A3) contributed
     by Minseo Lee.  Generated with his modifications of three
     combining patterns in the johab2ucs2 Perl script.
   * Modified U+5829 and U+6FF9 Chinese glyphs.
   * Full details in ChangeLog file.

2023-05-28: Release 15.0.04
   * Adjustments to Greek and Coptic letters.
   * Adjustments to two heart glyphs.
   * Added Engsvanáli (U+E100..U+E14F) CSUR glyphs.

2023-05-21: Release 15.0.03
   * Fixed a typo in src/Makefile that could affect
     "make install".
   * Install unifont_all.hex in $(PKGDEST) directory.
   * Other minor changes; see ChangeLog file for details.

2023-05-20: Release 15.0.02
   * N.B.: This is the last planned version to build TrueType
     font files as part of the standard font build with

           make BUILDFONT=1

     In the future, only OpenType font files will be part of
     the ordinary font building using the hex2otf program.
     TrueType fonts will have to be built manually using

           cd font
           make truetype

   * Minor update to allow for separate building of TrueType
     and OpenType fonts.  This is in preparation for the future
     removal of TrueType font files from default font builds,
     in preference for OpenType font files.  That change is
     planned for the release of Unifont 15.1.01, which will
     correspond to the release of the Unicode 15.1.0 standard
     scheduled for September 2023.
   * Added Doxygen annotation to C programs and corresponding
     ".h" header files.  Annotated output is generated in a
     new "doxygen" subdirectory.  Doxygen output consists of
     a browsable set of HTML pages and a hyperlinked PDF document.
   * Updates to make files to fix issues with redundant building
     under some circumstances.
   * Added 196 Plane 2 and Plane 3 Chinese ideographs drawn by
     湖 远星 to the basic Unifont font files for coverage of
     the General Standard Chinese Characters set.  Also added
     biang glyphs drawn by Ming Fan.
   * Fullwidth CJK characters that appeared to be boldface in
     the range U+FF00..U+FFEF were updated using glyphs from
     Wen Quan Yi, with some modifications.

2022-09-13: Release 15.0.01
   * Update for Unicode 15.0.0 release.
   * "make install" now installs Unifont OpenType font files in the
     same directory as TrueType fonts by default.  The intention is
     to remove TrueType fonts file from the "install" target for the
     Unicode 16.0.0 release unless there is a reason to keep them.
   * Existing Unicode scripts in Unifont maintain complete coverage.
   * Under ConScript Unicode Registry (UCSUR) scripts added:
     - U+E1D0..U+E1FF Syai
     - U+E360..U+E37F Sarkai
     - U+ED80..U+EDAF Iranic.
   * New Unicode 15.0.0 Plane 1 scripts:
     - U+10EFD..U+10EFF Arabic Extended-C
     - U+11B00..U+11B5F Devanagari Extended-A
     - U+11F00..U+11F5F Kawi
     - U+1D2C0..U+1D2DF Kaktovik Numerals
     - U+1E030..U+1E08F Cyrillic Exntended-D
     - U+1E4D0..U+1E4FF Nag Mundari.

2022-06-04: Release 14.0.04

src/unipagecount.c
/**
   @file unipagecount.c

   @brief unipagecount - Count the number of glyphs defined in each page
                         of 256 code points

   @author Paul Hardy, unifoundry <at> unifoundry.com, December 2007
   
   @copyright Copyright (C) 2007, 2008, 2013, 2014 Paul Hardy

   This program counts the number of glyphs that are defined in each
   "page" of 256 code points, and prints the counts in an 8 x 8 grid.
   Input is from stdin.  Output is to stdout.

   The background color of each cell in a 16-by-16 grid of 256 code points
   is shaded to indicate percentage coverage.  Red indicates 0% coverage,
   green represents 100% coverage, and colors in between pure red and pure
   green indicate partial coverage on a scale.

   Each code point range number can be a hyperlink to a PNG file for
   that 256-code point range's corresponding bitmap glyph image.

   Synopsis:

             unipagecount < font_file.hex > count.txt
             unipagecount -phex_page_num < font_file.hex  -- just 256 points
             unipagecount -h < font_file.hex              -- HTML table
             unipagecount -P1 -h < font.hex > count.html  -- Plane 1, HTML out
             unipagecount -l < font_file.hex              -- linked HTML table
*/
/*
   LICENSE:

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 2 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
   2018, Paul Hardy: Changed "Private Use" to "Private Use Area" in
   output HTML file.

   21 October 2023 [Paul Hardy]:
   - Added full prototype for mkftable function in main function.
*/

#include <stdio.h>
#include <stdlib.h>

#define MAXBUF 256 ///< Maximum input line size - 1.


/**
   @brief The main function.

   @param[in] argc The count of command line arguments.
   @param[in] argv Pointer to array of command line arguments.
   @return This program exits with status 0.
*/
int
main (int argc, char *argv[])
{

   char inbuf[MAXBUF]; /* Max 256 characters in an input line */
   int i, j;  /* loop variables */
   unsigned plane=0; /* Unicode plane number, 0 to 0x16 */
   unsigned page;  /* unicode page (256 bytes wide) */
   unsigned unichar; /* unicode character */
   int pagecount[256] = {256 * 0};
   int onepage=0; /* set to one if printing character grid for one page */
   int pageno=0; /* page number selected if only examining one page */
   int html=0;   /* =0: print plain text; =1: print HTML */
   int links=0;  /* =1: print HTML links; =0: don't print links */

   /* make (print) flipped HTML table */
   void mkftable (unsigned plane, int pagecount[256], int links);

   size_t strlen();

   if (argc > 1 && argv[1][0] == '-') {  /* Parse option */
      plane = 0;
      for (i = 1; i < argc; i++) {
         switch (argv[i][1]) {
            case 'p':  /* specified -p<hexpage> -- use given page number */
               sscanf (&argv[1][2], "%x", &pageno);
               if (pageno >= 0 && pageno <= 255) onepage = 1;
               break;
            case 'h':  /* print HTML table instead of text table */
               html = 1;
               break;
            case 'l':  /* print hyperlinks in HTML table */
               links = 1;
               html = 1;
               break;
            case 'P':  /* Plane number specified */
               plane = atoi(&argv[1][2]);
               break;
         }
      }
   }
   /*
      Initialize pagecount to account for noncharacters.
   */
   if (!onepage && plane==0) {
      pagecount[0xfd] = 32;  /* for U+FDD0..U+FDEF */
   }
   pagecount[0xff] = 2;   /* for U+nnFFFE, U+nnFFFF */
   /*
      Read one line at a time from input.  The format is:

         <hexpos>:<hexbitmap>

      where <hexpos> is the hexadecimal Unicode character position
      in the range 00..FF and <hexbitmap> is the sequence of hexadecimal
      digits of the character, laid out in a grid from left to right,
      top to bottom.  The character is assumed to be 16 rows of variable
      width.
   */
   while (fgets (inbuf, MAXBUF-1, stdin) != NULL) {
      sscanf (inbuf, "%X", &unichar);
      page = unichar >> 8;
      if (onepage) { /* only increment counter if this is page we want */
         if (page == pageno) { /* character is in the page we want */
            pagecount[unichar & 0xff]++; /* mark character as covered */
         }
      }
      else { /* counting all characters in all pages */
         if (plane == 0) {
            /* Don't add in noncharacters (U+FDD0..U+FDEF, U+FFFE, U+FFFF) */
            if (unichar < 0xfdd0 || (unichar > 0xfdef && unichar < 0xfffe))
               pagecount[page]++;
         }
         else {
            if ((page >> 8) == plane) { /* code point is in desired plane */
               pagecount[page & 0xFF]++;
            }
         }
      }
   }
   if (html) {
      mkftable (plane, pagecount, links);
   }
   else {  /* Otherwise, print plain text table */
      if (plane > 0) fprintf (stdout, "  ");
      fprintf (stdout,
         "     0   1   2   3   4   5   6   7   8   9   A   B   C   D   E   F\n");
      for (i=0; i<0x10; i++) {
         fprintf (stdout,"%02X%X ", plane, i); /* row header */
         for (j=0; j<0x10; j++) {
            if (onepage) {
               if (pagecount[i*16+j])
                  fprintf (stdout," *  ");
               else
                  fprintf (stdout," .  ");
            }
            else {
               fprintf (stdout, "%3X ", pagecount[i*16+j]);
            }
         }
         fprintf (stdout,"\n");
      }
   
   }
   exit (0);
}


/**
   @brief Create an HTML table linked to PNG images.

   This function creates an HTML table to show PNG files
   in a 16 by 16 grid.  The background color of each "page"
   of 256 code points is shaded from red (for 0% coverage)
   to green (for 100% coverage).

   @param[in] plane The Unicode plane, 0..17.
   @param[in] pagecount Array with count of glyphs in each 256 code point range.
   @param[in] links 1 = generate hyperlinks, 0 = do not generate hyperlinks.
*/
void
mkftable (unsigned plane, int pagecount[256], int links)
{
   int i, j;
   int count;
   unsigned bgcolor;
   
   printf ("<html>\n");
   printf ("<body>\n");
   printf ("<table border=\"3\" align=\"center\">\n");
   printf ("  <tr><th colspan=\"16\" bgcolor=\"#ffcc80\">");
   printf ("GNU Unifont Glyphs<br>with Page Coverage for Plane %d<br>(Green=100%%, Red=0%%)</th></tr>\n", plane);
   for (i = 0x0; i <= 0xF; i++) {
      printf ("  <tr>\n");
      for (j = 0x0; j <= 0xF; j++) {
         count = pagecount[ (i << 4) | j ];
         
         /* print link in cell if links == 1 */
         if (plane != 0 || (i < 0xd || (i == 0xd && j < 0x8) || (i == 0xf && j > 0x8))) {
            /* background color is light green if completely done */
            if (count == 0x100) bgcolor = 0xccffcc;
            /* otherwise background is a shade of yellow to orange to red */
            else bgcolor = 0xff0000 | (count << 8) | (count >> 1);
            printf ("    <td bgcolor=\"#%06X\">", bgcolor);
            if (plane == 0)
               printf ("<a href=\"png/plane%02X/uni%02X%X%X.png\">%X%X</a>", plane, plane, i, j, i, j);
            else
               printf ("<a href=\"png/plane%02X/uni%02X%X%X.png\">%02X%X%X</a>", plane, plane, i, j, plane, i, j);
            printf ("</td>\n");
         }
         else if (i == 0xd) {
            if (j == 0x8) {
               printf ("    <td align=\"center\" colspan=\"8\" bgcolor=\"#cccccc\">");
               printf ("<b>Surrogate Pairs</b>");
               printf ("</td>\n");
            }  /* otherwise don't print anything more columns in this row */
         }
         else if (i == 0xe) {
            if (j == 0x0) {
               printf ("    <td align=\"center\" colspan=\"16\" bgcolor=\"#cccccc\">");
               printf ("<b>Private Use Area</b>");
               printf ("</td>\n");
            }  /* otherwise don't print any more columns in this row */
         }
         else if (i == 0xf) {
            if (j == 0x0) {
               printf ("    <td align=\"center\" colspan=\"9\" bgcolor=\"#cccccc\">");
               printf ("<b>Private Use Area</b>");
               printf ("</td>\n");
            }
         }
      }
      printf ("  </tr>\n");
   }
   printf ("</table>\n");
   printf ("</body>\n");
   printf ("</html>\n");

   return;
}
